Plugin does not work with IntelliJ 2020.2

(copied from review at https://plugins.jetbrains.com/plugin/7908-fitnesse by Mathias Rühl)

FitNesse Plugin 1.10.1 leads to exceptions in IntelliJ 2020.2 with FitNesse Version "fitnesse-standalone-20190716.jar".
I had to disable/uninstall that Plugin to have a working IDE, as IntelliJ does not behave correctly when the Exception occur.

```
Error while indexing /user/home/delevelop/FitNesseRoot/FitNesse/UserGuide/content.txt
To reindex this file IDEA has to be restarted

com.intellij.util.indexing.impl.MapReduceIndex$MapInputException: Failed to map data for input 15878
	at com.intellij.util.indexing.impl.MapReduceIndex.mapByIndexer(MapReduceIndex.java:300)
	at com.intellij.util.indexing.impl.MapReduceIndex.mapInput(MapReduceIndex.java:285)
	at com.intellij.util.indexing.impl.storage.VfsAwareMapReduceIndex.mapInput(VfsAwareMapReduceIndex.java:157)
	at com.intellij.util.indexing.impl.storage.VfsAwareMapReduceIndex.mapInput(VfsAwareMapReduceIndex.java:41)
	at com.intellij.util.indexing.impl.MapReduceIndex.mapInputAndPrepareUpdate(MapReduceIndex.java:223)
	at com.intellij.util.indexing.FileBasedIndexImpl.updateSingleIndex(FileBasedIndexImpl.java:1387)
	at com.intellij.util.indexing.FileBasedIndexImpl.lambda$doIndexFileContent$25(FileBasedIndexImpl.java:1292)
	at com.intellij.openapi.fileTypes.impl.FileTypeManagerImpl.freezeFileTypeTemporarilyIn(FileTypeManagerImpl.java:557)
	at com.intellij.util.indexing.FileBasedIndexImpl.doIndexFileContent(FileBasedIndexImpl.java:1262)
	at com.intellij.util.indexing.FileBasedIndexImpl.indexFileContent(FileBasedIndexImpl.java:1214)
	at com.intellij.util.indexing.contentQueue.IndexUpdateRunner.lambda$indexOneFileOfJob$1(IndexUpdateRunner.java:195)
	at com.intellij.openapi.application.impl.NonBlockingReadActionImpl$Submission.insideReadAction(NonBlockingReadActionImpl.java:498)
	at com.intellij.openapi.application.impl.NonBlockingReadActionImpl$Submission.lambda$attemptComputation$2(NonBlockingReadActionImpl.java:446)
	at com.intellij.openapi.application.impl.ApplicationImpl.tryRunReadAction(ApplicationImpl.java:1110)
	at com.intellij.openapi.progress.util.ProgressIndicatorUtils.lambda$runInReadActionWithWriteActionPriority$0(ProgressIndicatorUtils.java:77)
	at com.intellij.openapi.progress.util.ProgressIndicatorUtils.runActionAndCancelBeforeWrite(ProgressIndicatorUtils.java:153)
	at com.intellij.openapi.progress.util.ProgressIndicatorUtils.lambda$runWithWriteActionPriority$1(ProgressIndicatorUtils.java:114)
	at com.intellij.openapi.progress.ProgressManager.lambda$runProcess$0(ProgressManager.java:56)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$2(CoreProgressManager.java:170)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:629)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:581)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:60)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:157)
	at com.intellij.openapi.progress.ProgressManager.runProcess(ProgressManager.java:56)
	at com.intellij.openapi.progress.util.ProgressIndicatorUtils.runWithWriteActionPriority(ProgressIndicatorUtils.java:111)
	at com.intellij.openapi.progress.util.ProgressIndicatorUtils.runInReadActionWithWriteActionPriority(ProgressIndicatorUtils.java:77)
	at com.intellij.openapi.application.impl.NonBlockingReadActionImpl$Submission.attemptComputation(NonBlockingReadActionImpl.java:463)
	at com.intellij.openapi.application.impl.NonBlockingReadActionImpl$Submission.executeSynchronously(NonBlockingReadActionImpl.java:401)
	at com.intellij.openapi.application.impl.NonBlockingReadActionImpl.executeSynchronously(NonBlockingReadActionImpl.java:181)
	at com.intellij.util.indexing.contentQueue.IndexUpdateRunner.indexOneFileOfJob(IndexUpdateRunner.java:198)
	at com.intellij.util.indexing.contentQueue.IndexUpdateRunner.indexJobsFairly(IndexUpdateRunner.java:152)
	at com.intellij.util.indexing.contentQueue.IndexUpdateRunner.lambda$indexFiles$0(IndexUpdateRunner.java:98)
	at com.intellij.util.concurrency.BoundedTaskExecutor.doRun(BoundedTaskExecutor.java:215)
	at com.intellij.util.concurrency.BoundedTaskExecutor.access$200(BoundedTaskExecutor.java:26)
	at com.intellij.util.concurrency.BoundedTaskExecutor$1.execute(BoundedTaskExecutor.java:194)
	at com.intellij.util.ConcurrencyUtil.runUnderThreadName(ConcurrencyUtil.java:207)
	at com.intellij.util.concurrency.BoundedTaskExecutor$1.run(BoundedTaskExecutor.java:183)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:668)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:665)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1.run(Executors.java:665)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: com.intellij.diagnostic.PluginException: Discontinuous sequence of tokens is generated by lexer: fitnesse.idea.lexer.FitnesseLexer [Plugin: FitNesse]
	at com.intellij.ide.plugins.PluginManagerCore.createPluginException(PluginManagerCore.java:290)
	at com.intellij.diagnostic.PluginProblemReporterImpl.createPluginExceptionByClass(PluginProblemReporterImpl.java:12)
	at com.intellij.diagnostic.PluginException.createByClass(PluginException.java:58)
	at com.intellij.openapi.editor.ex.util.ValidatingLexerWrapper.throwException(ValidatingLexerWrapper.java:126)
	at com.intellij.openapi.editor.ex.util.ValidatingLexerWrapper.advance(ValidatingLexerWrapper.java:94)
	at com.intellij.openapi.editor.ex.util.LexerEditorHighlighter.doSetText(LexerEditorHighlighter.java:441)
	at com.intellij.openapi.editor.ex.util.LexerEditorHighlighter.setText(LexerEditorHighlighter.java:409)
	at com.intellij.psi.impl.cache.impl.id.PlatformIdTableBuilding$TokenSetTodoIndexer.map(PlatformIdTableBuilding.java:104)
	at com.intellij.psi.impl.cache.impl.id.PlatformIdTableBuilding$TokenSetTodoIndexer.map(PlatformIdTableBuilding.java:83)
	at com.intellij.psi.impl.cache.impl.todo.TodoIndex$4.map(TodoIndex.java:100)
	at com.intellij.psi.impl.cache.impl.todo.TodoIndex$4.map(TodoIndex.java:77)
	at com.intellij.util.indexing.CompositeDataIndexer.map(CompositeDataIndexer.java:54)
	at com.intellij.util.indexing.CompositeDataIndexer.map(CompositeDataIndexer.java:18)
	at com.intellij.util.indexing.impl.MapReduceIndex.mapByIndexer(MapReduceIndex.java:294)
	... 43 more
```

--- a/src/main/resources/META-INF/plugin.xml
+++ b/src/main/resources/META-INF/plugin.xml
@@ -8,6 +8,7 @@
 
 1.11.0
 
+ * Fixed parsing of wiki files to ensure continous tokens, required for IntelliJ 2020.2. See https://github.com/gshakhn/idea-fitnesse/issues/55.
  * Added plugin logo (https://github.com/gshakhn/idea-fitnesse/issues/51).
 
 1.10.1

--- a/src/main/scala/fitnesse/idea/lexer/FitnesseLexer.scala
+++ b/src/main/scala/fitnesse/idea/lexer/FitnesseLexer.scala
@@ -44,6 +44,7 @@ class FitnesseLexer extends LexerBase {
     symbolList = fetchNextSymbols()
 
     // if TABLE_START, ROW_START -> reduce to TABLE_START, fix offsets
+    // if COLLAPSIBLE_START, title ->  fix offsets
     // drop all CELL_START | ROW_START -> advance
     // CELL_END, ROW_END -> advance to ROW_END
     // ROW_END, TABLE_END -> advance to TABLE_END
@@ -61,6 +62,12 @@ class FitnesseLexer extends LexerBase {
         case table: Table            =>
           // Fetch new simples to ensure the table contents are properly parsed.
           symbolList = LexerSymbol(FitnesseTokenType.TABLE_START, symbol.getStartOffset, symbol.getChildren.get(0).getStartOffset) :: fetchNextSymbols()
+        case _: Collapsible            =>
+          // Fetch new simples to ensure the collapsible contents are properly parsed.
+          val title :: tail = fetchNextSymbols()
+          val titleEnd = title.getChildren.last.getEndOffset
+          val startAfter = previousEnd(symbol, titleEnd, title.getEndOffset)
+          symbolList = LexerSymbol(FitnesseTokenType.COLLAPSIBLE_START, symbol.getStartOffset, title.getStartOffset) :: title :: LexerSymbol(FitnesseTokenType.LINE_TERMINATOR, startAfter, startAfter + 1) :: tail
         case _ =>
       }
       case _ =>
@@ -139,13 +146,20 @@ object FitnesseLexer {
   case class LexerSymbol(elementType: IElementType, start: Int, end: Int) extends Symbol(new SymbolType(elementType.toString), "", start, end)
 
   def terminatorFor(symbol: Symbol): Option[Symbol] = {
-    val startOffset = if (lastChild(symbol).getEndOffset == symbol.getEndOffset) lastChild(symbol).getStartOffset else lastChild(symbol).getEndOffset
+    val lastNestedChild : Symbol = lastChild(symbol)
+    val symbolEndOffset = symbol.getEndOffset
+    val startOffset = if (lastNestedChild.getEndOffset == symbolEndOffset) lastNestedChild.getStartOffset else lastNestedChild.getEndOffset
+    val previousEndOffset = previousEnd(symbol, startOffset, symbolEndOffset)
 
     symbol.getType match {
-      case _ : Table                 => Some(LexerSymbol(FitnesseTokenType.TABLE_END, startOffset, symbol.getEndOffset))
-      case _ : Collapsible           => Some(LexerSymbol(FitnesseTokenType.COLLAPSIBLE_END, lastChild(symbol).getEndOffset, symbol.getEndOffset))
-      case s if s eq Table.tableRow  => Some(LexerSymbol(FitnesseTokenType.ROW_END, startOffset, symbol.getEndOffset))
-      case s if s eq Table.tableCell => Some(LexerSymbol(FitnesseTokenType.CELL_END, startOffset, symbol.getEndOffset))
+      case _ : Table                 => Some(LexerSymbol(FitnesseTokenType.TABLE_END, previousEndOffset, symbolEndOffset))
+      case _ : Collapsible           => {
+        val childPreviousEndOffset = previousEnd(symbol.getChildren.last, startOffset, symbol.getChildren.last.getEndOffset)
+        val collEndStart = if (symbolEndOffset - childPreviousEndOffset > 1) childPreviousEndOffset else lastNestedChild.getEndOffset
+        Some(LexerSymbol(FitnesseTokenType.COLLAPSIBLE_END, collEndStart, symbolEndOffset))
+      }
+      case s if s eq Table.tableRow  => Some(LexerSymbol(FitnesseTokenType.ROW_END, previousEndOffset, symbolEndOffset))
+      case s if s eq Table.tableCell => Some(LexerSymbol(FitnesseTokenType.CELL_END, previousEndOffset, symbolEndOffset))
       case _ => None
     }
   }
@@ -153,6 +167,13 @@ object FitnesseLexer {
   private def lastChild(symbol: Symbol): Symbol = {
     val children = symbol.getChildren
     if (children.isEmpty || children.exists(_.getStartOffset == -1)) symbol else lastChild(children.last)
+  }
+
+  private def previousEnd(symbol: Symbol, previousMax: Int, endOffset: Int): Int = {
+    val children = symbol.getChildren
+    val symbolEnd = symbol.getEndOffset
+    val currentEnd = if ((symbolEnd < endOffset) && (symbolEnd > previousMax)) symbolEnd else previousMax
+    if (children.isEmpty || children.exists(_.getStartOffset == -1)) currentEnd else previousEnd(children.last, currentEnd, endOffset)
   }
 }
 

--- a/src/test/scala/fitnesse/idea/lexer/MiscLexerSuite.scala
+++ b/src/test/scala/fitnesse/idea/lexer/MiscLexerSuite.scala
@@ -107,8 +107,9 @@ class MiscLexerSuite extends LexerSuite {
   test("Collapsible section") {
     assertResult(
       List(
-        (FitnesseTokenType.COLLAPSIBLE_START, "!* abc\ndef\n*!"),
+        (FitnesseTokenType.COLLAPSIBLE_START, "!* "),
         (FitnesseTokenType.WORD, "abc"),
+        (FitnesseTokenType.LINE_TERMINATOR, "\n"),
         (FitnesseTokenType.WORD, "def"),
         (FitnesseTokenType.LINE_TERMINATOR, "\n"),
         (FitnesseTokenType.COLLAPSIBLE_END, "*!"),
@@ -119,19 +120,71 @@ class MiscLexerSuite extends LexerSuite {
     }
   }
 
+  test("Collapsed section") {
+    assertResult(
+      List(
+        (FitnesseTokenType.COLLAPSIBLE_START, 0, 7, "!****< "),
+        (FitnesseTokenType.WORD, 7, 10, "bbc"),
+        (FitnesseTokenType.LINE_TERMINATOR, 10, 11, "\n"),
+        (FitnesseTokenType.WORD, 11, 14, "def"),
+        (FitnesseTokenType.LINE_TERMINATOR, 14, 15, "\n"),
+        (FitnesseTokenType.COLLAPSIBLE_END, 15, 17, "*!"),
+        (FitnesseTokenType.WHITE_SPACE, 17, 18, " "),
+        (FitnesseTokenType.WORD, 18, 22, "word")
+      )) {
+      lexWithOffset("!****< bbc\ndef\n*! word")
+    }
+  }
+
+  test("Collapsible section multi word title") {
+    assertResult(
+      List(
+        (FitnesseTokenType.COLLAPSIBLE_START, 0, 4, "!*> "),
+        (FitnesseTokenType.WORD, 4, 9, "Extra"),
+        (FitnesseTokenType.WHITE_SPACE, 9, 10, " "),
+        (FitnesseTokenType.WORD, 10, 16, "Import"),
+        (FitnesseTokenType.LINE_TERMINATOR, 16, 17, "\n"),
+        (FitnesseTokenType.LINE_TERMINATOR, 17, 18, "\n"),
+        (FitnesseTokenType.WORD, 18, 20, "Hi"),
+        (FitnesseTokenType.LINE_TERMINATOR, 20, 21, "\n"),
+        (FitnesseTokenType.COLLAPSIBLE_END, 21, 24, "*!\n"),
+        (FitnesseTokenType.LINE_TERMINATOR, 24, 25, "\n")
+      )) {
+      lexWithOffset("!*> Extra Import\n\nHi\n*!\n\n")
+    }
+  }
+
   test("Collapsible section containing a table") {
     assertResult(
       List(
-        (FitnesseTokenType.COLLAPSIBLE_START, "!* title\n|abc|\n|xyz|\n*!"),
-        (FitnesseTokenType.WORD, "title"),
-        (FitnesseTokenType.TABLE_START, "|"),
-        (FitnesseTokenType.WORD, "abc"),
-        (FitnesseTokenType.ROW_END, "|\n|"),
-        (FitnesseTokenType.WORD, "xyz"),
-        (FitnesseTokenType.TABLE_END, "|\n"),
-        (FitnesseTokenType.COLLAPSIBLE_END, "|\n*!")
+        (FitnesseTokenType.COLLAPSIBLE_START, 0, 3, "!* "),
+        (FitnesseTokenType.WORD, 3, 8, "title"),
+        (FitnesseTokenType.LINE_TERMINATOR, 8, 9, "\n"),
+        (FitnesseTokenType.TABLE_START, 9, 10, "|"),
+        (FitnesseTokenType.WORD, 10, 13, "abc"),
+        (FitnesseTokenType.ROW_END, 13, 16, "|\n|"),
+        (FitnesseTokenType.WORD, 16, 19, "xyz"),
+        (FitnesseTokenType.TABLE_END, 19, 21, "|\n"),
+        (FitnesseTokenType.COLLAPSIBLE_END, 21, 23, "*!")
       )) {
-      lex("!* title\n|abc|\n|xyz|\n*!")
+      lexWithOffset("!* title\n|abc|\n|xyz|\n*!")
+    }
+  }
+
+  test("Collapsible section containing a table followed by newline") {
+    assertResult(
+      List(
+        (FitnesseTokenType.COLLAPSIBLE_START, 0, 3, "!* "),
+        (FitnesseTokenType.WORD, 3, 8, "title"),
+        (FitnesseTokenType.LINE_TERMINATOR, 8, 9, "\n"),
+        (FitnesseTokenType.TABLE_START, 9, 10, "|"),
+        (FitnesseTokenType.WORD, 10, 13, "abc"),
+        (FitnesseTokenType.ROW_END, 13, 16, "|\n|"),
+        (FitnesseTokenType.WORD, 16, 19, "xyz"),
+        (FitnesseTokenType.TABLE_END, 19, 21, "|\n"),
+        (FitnesseTokenType.COLLAPSIBLE_END, 21, 24, "*!\n")
+      )) {
+      lexWithOffset("!* title\n|abc|\n|xyz|\n*!\n")
     }
   }
 

--- a/src/test/scala/fitnesse/idea/lexer/TableLexerSuite.scala
+++ b/src/test/scala/fitnesse/idea/lexer/TableLexerSuite.scala
@@ -315,4 +315,51 @@ class TableLexerSuite extends LexerSuite {
       lexWithOffset("!||\n||")
     }
   }
+
+  test("Formatted cells should not make table offsets overlap") {
+    assertResult(
+      List(
+        (FitnesseTokenType.TABLE_START, 0, 1, "|"),
+        (FitnesseTokenType.BOLD,        1, 8, "'''E'''"),
+        (FitnesseTokenType.CELL_END,    8, 9, "|"),
+        (FitnesseTokenType.ITALIC,      9, 14, "''A''"),
+        (FitnesseTokenType.TABLE_END,   14, 15, "|")
+      )) {
+      lexWithOffset("|'''E'''|''A''|")
+    }
+  }
+
+  test("Formatted cells should not make table offsets overlap with multiple rows") {
+    assertResult(
+      List(
+        (FitnesseTokenType.TABLE_START, 0, 1, "|"),
+        (FitnesseTokenType.BOLD,        1, 8, "'''E'''"),
+        (FitnesseTokenType.CELL_END,    8, 9, "|"),
+        (FitnesseTokenType.ITALIC,      9, 14, "''A''"),
+        (FitnesseTokenType.ROW_END,     14, 17, "|\n|"),
+        (FitnesseTokenType.BOLD,        17, 24, "'''E'''"),
+        (FitnesseTokenType.CELL_END,    24, 25, "|"),
+        (FitnesseTokenType.ITALIC,      25, 30, "''A''"),
+        (FitnesseTokenType.TABLE_END,   30, 31, "|")
+      )) {
+      lexWithOffset("|'''E'''|''A''|\n|'''E'''|''A''|")
+    }
+  }
+
+  test("Formatted cells should not make table offsets overlap with multiple rows and text after") {
+    assertResult(
+      List(
+        (FitnesseTokenType.TABLE_START, 0, 1, "|"),
+        (FitnesseTokenType.BOLD,        1, 8, "'''E'''"),
+        (FitnesseTokenType.CELL_END,    8, 9, "|"),
+        (FitnesseTokenType.ITALIC,      9, 14, "''A''"),
+        (FitnesseTokenType.ROW_END,     14, 17, "|\n|"),
+        (FitnesseTokenType.BOLD,        17, 24, "'''E'''"),
+        (FitnesseTokenType.CELL_END,    24, 25, "|"),
+        (FitnesseTokenType.ITALIC,      25, 30, "''A''"),
+        (FitnesseTokenType.TABLE_END,   30, 32, "|\n")
+      )) {
+      lexWithOffset("|'''E'''|''A''|\n|'''E'''|''A''|\n")
+    }
+  }
 }

--- a/src/test/scala/fitnesse/idea/parser/MiscParserSuite.scala
+++ b/src/test/scala/fitnesse/idea/parser/MiscParserSuite.scala
@@ -49,16 +49,18 @@ class MiscParserSuite extends ParserSuite {
       Node(FitnesseElementType.FILE, List(
         Node(FitnesseElementType.COLLAPSIBLE, List(
           Leaf(FitnesseTokenType.COLLAPSIBLE_START, "!* "),
-          Leaf(FitnesseTokenType.WORD, "title\n"),
+          Leaf(FitnesseTokenType.WORD, "title"),
+          Leaf(FitnesseTokenType.WHITE_SPACE, "\n"),
           Node(TableElementType.DECISION_TABLE, List(
             Leaf(FitnesseTokenType.TABLE_START,"|"),
             Node(FitnesseElementType.ROW, List(
               Node(FitnesseElementType.FIXTURE_CLASS, List(
                 Leaf(FitnesseTokenType.WORD, "abc")
               ))
-            ))
+            )),
+            Leaf(FitnesseTokenType.TABLE_END,"|\n")
           )),
-          Leaf(FitnesseTokenType.COLLAPSIBLE_END, "|\n*!\n")
+          Leaf(FitnesseTokenType.COLLAPSIBLE_END, "*!\n")
         ))
       ))
     ) {

